src/Chapter2.hs

```diff
@@ -621,7 +641,14 @@ Write a function that takes elements of a list only on even positions.
 >>> takeEven [2, 1, 3, 5, 4]
 [2,3,4]
 -}
-takeEven = error "takeEven: Not implemented!"
+takeEven :: [Int] -> [Int]
+takeEven lst = go [] $ zip [0..] lst
+  where
+    go :: [Int] -> [(Int, Int)] -> [Int]
+    go res [] = reverse res
+    go res ((el, x):xs)
+      | el `mod` 2 == 0 = go (x : res) xs
+      | otherwise       = go res xs
```

Comment:
Nice idea on the helper functions :+1:

A few additional thoughts on your idea. Instead of creating a list of integers, you can have a single value representing current index and increment it on each step. So you will allocate less memory :slightly_smiling_face: 

After this implementation you can notice, that as you have only two states of the operation (odd or even), instead of accumulators of the current position number, you can just use `Bool`. And on each next recursion call, you just give it `not acc`. This will remove the necessity to check `mod 2` each time, which would save a couple of functions call on each interaction.

Moreover, you do `reverse` in the end, and this is a bit slow. If you append elements directly to the results (e.g. `x : go xs`) instead of to the beginning of `res`, you will get the correct order of elements immediately.

Alternatively, you could use pattern matching in `takeEven` instead. As you need to take every second element, it is possible to represent with a few pattern matching clauses 🙂

Reply:
Thanks! It's really nice to know what approach are efficient and what are not